File: algorithms/Java/quicksort.java
class Solution
{
    static void quickSort(int arr[], int low, int high)
    {
        if ( low > high) {
            return;
        }
        
        int pivot = partition(arr, low, high);
        quickSort(arr, low, pivot-1);
        quickSort(arr, pivot+1, high);
    }
    
    static int partition(int arr[], int low, int high)
    {
        int pivot = low;
        for (int i = low; i<high; i++) {
            if (arr[i] < arr[high]) {
                if (pivot != i) {
                    int temp = arr[i];
                    arr[i] = arr[pivot];
                    arr[pivot] = temp;
                }
                pivot += 1;
            } 
        }
        
        //put pivot element to it's place
       if (pivot < high) {
            int temp = arr[high];
            arr[high] = arr[pivot];
            arr[pivot] = temp;
        }

        return pivot;
    } 
}

File: Swift/quickSort.swift

extension Array where Element: Comparable {
    public mutating func quickSort() {
        quickSort(low: 0, high: count-1)
    }
    
    private mutating func quickSort(low: Int, high: Int) {
        
        guard low <= high else { return}
        
        let pivot = partition(low: low, high: high)
        quickSort(low: low, high: pivot-1)
        quickSort(low: pivot+1, high: high)
    }
    
    private mutating func partition(low: Int, high: Int) -> Int {
        var pivot = low
        
        for i in low..<high {
            if self[i] < self[high] {
                if pivot != i {
                    swapElement(at: i, withElementAt: pivot)
                }
                pivot += 1
            }
        }
        
        if pivot < high {
            swapElement(at: pivot, withElementAt: high)
        }
        return pivot
    }
    
    private mutating func swapElement(at i: Int, withElementAt j: Int) {
        let temp = self[i]
        self[i] = self[j]
        self[j] = temp
    }
}

//test 
var array = [2, 1, 6, 10, 4, 1, 3, 9, 7]
array.quickSort()
print(array)
//output: [1, 1, 2, 3, 4, 6, 7, 9, 10]
